crypto/ctype.c

```diff
@@ -278,3 +289,66 @@ int ossl_ascii_isdigit(const char inchar) {
         return 1;
     return 0;
 }
+
+/* str[n]casecmp_l is defined in POSIX 2008-01. Value is taken accordingly
+ * https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html */
+
+#if (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L)
+static locale_t loc;
+
+# ifndef FIPS_MODULE
+void *OPENSSL_strcasecmp_locale() {
+    return (void *)loc;
+}
+# endif
+
+int ossl_init_casecmp_int() {
+    loc = newlocale(LC_COLLATE_MASK, "C", (locale_t) 0);
+    return (loc == (locale_t) 0) ? 0 : 1;
+}
+
+void ossl_deinit_casecmp() {
+    freelocale(loc);
+}
+
+int ossl_strcasecmp(const char *s1, const char *s2)
+{
+    return strcasecmp_l(s1, s2, (locale_t)OPENSSL_strcasecmp_locale());
+}
+
+int ossl_strncasecmp(const char *s1, const char *s2, size_t n)
+{
+    return strncasecmp_l(s1, s2, n, (locale_t)OPENSSL_strcasecmp_locale());
+}
+#else
+# ifndef FIPS_MODULE
+void *OPENSSL_strcasecmp_locale() {
```

Comment:
What happens inside the FIPS provider?  Where does this function come from?

It isn't allowed to come from libcrypto.

Reply:
I was passing it on the initialization stage. If it is wrong, I can try reimplement it in fips provider itself, though I am not sure where to free the locale object

Reply:
But what about the ossl_strcasecmp itself?

Reply:
Every call from the FIPS provider back to libcrypto has to be done via a callback.

Or we can include the code in the FIPS provider -- I think this might be the better alternative.

Wouldn't `fips_intern_teardown()` be the spot to free the locale?

Reply:
OK, I will try to implement it in FIPS provider itself.

Reply:
Looks like it's OK when implemented in the fips provider, thanks.